### PR TITLE
[C10D] Add __repr__ to P2POp class

### PR DESCRIPTION
**Pull request by wconstab**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #126582
* #126539
* __->__ #126538
* #126419

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @yf225 @chauhang @d4l3k

### CODE REVIEW

**Comment by wconstab** [comment]
Wonder whether folks prefer local ranks or global ranks for src/dsc

I went with local since i thought its helpful to print the PG too and if you print the PG and then print global ranks its confusing.  Also, the log-prefix for each log line usually shows the global rank, so you get both pieces of info.  And local rank is more 'helpful' in some cases (stage 0 is local rank 0, for PP)

otoh our APIs are confusing, and they ask you to specify a PG and also specify global ranks, so maybe printing global ranks is most consistent.

**Review by Skylion007 — APPROVED** [review]

**Review by kwen2501 — APPROVED** [review]
LGTM. Thanks for adding it!

File: torch/distributed/distributed_c10d.py [hunk 1]
```diff
@@ -408,6 +408,21 @@ def __new__(cls, op: Callable, tensor: torch.Tensor, peer: int,
...
+        if "send" in op_name:
+            s = my_group_rank
+            d = peer_group_rank
+        elif "recv" in op_name:
+            s = peer_group_rank
+            d = my_group_rank
+        else:
+            return super().__repr__()
+
+        return f"P2POp({op_name} pg={group_name}, s={s}, d={d},  {self.tensor.shape}, {self.tensor.dtype})"
```
Comment (c-p-i-o) [inline torch/distributed/distributed_c10d.py:425]: Print the `tag` too if it exists? 
Don't know how/when it's used - but might be useful?

**Review by c-p-i-o — APPROVED** [review]

File: torch/distributed/distributed_c10d.py (hunk 1 above)
Comment (wconstab) [reply torch/distributed/distributed_c10d.py:425]: yea.... actually NCCL doesn't even support tag.  So i would print it out of 'completeness' or compatibility with Gloo, but i think nobody is using tag anyway and we should probably deprecate it.

**Pull request closed without merging**